internal/api/provider/eip4361.go

```diff
@@ -0,0 +1,202 @@
+	return "" // Web3 auth doesn't use OAuth flow
+}
+
+func (p *Web3Provider) GetOAuthToken(code string) (*oauth2.Token, error) {
+	return nil, errors.New("GetOAuthToken not implemented for Web3")
+}
+
+func (p *Web3Provider) GetUserData(ctx context.Context, tok *oauth2.Token) (*UserProvidedData, error) {
+	return nil, errors.New("GetUserData not implemented for Web3")
+}
+
+// VerifySignedMessage verifies a signed Web3 message based on the blockchain
+func (p *Web3Provider) VerifySignedMessage(db *storage.Connection, params *Web3GrantParams) (*UserProvidedData, error) {   
+	var err error
+	
+	parsedMessage, err := siws.ParseSIWSMessage(params.Message)
+
+	if err != nil {
+		return nil, siws.ErrorMalformedMessage
+	}
+	
+	
+
+	// Verify and consume nonce first
+	if err := crypto.VerifyAndConsumeNonce(db, parsedMessage.Nonce, parsedMessage.Address); err != nil {
+		return nil, siws.ErrorCodeInvalidNonce
+	}
+	network := strings.Split(params.Chain, ":")
+	if len(network) != 2 {
+		return nil, siws.ErrInvalidChainID
+	}
+	chain := network[0]
+	if chain == "" {
+		return nil, siws.ErrInvalidChainID
+	}
+	
+	switch chain {
+	case BlockchainEthereum:
+		return nil, httpError(http.StatusNotImplemented, "signature verification not implemented for %s", network)
+		
+	case BlockchainSolana:
+		err = p.verifySolanaSignature(params.Signature, params.Message, parsedMessage)
+	default:
+		return nil, httpError(http.StatusNotImplemented, "signature verification not implemented for %s", network)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("signature verification failed: %w", err)
+	}
+
+	// Construct the provider_id as network:chain:address to make it unique
+	providerId := fmt.Sprintf("%s:%s", params.Chain, parsedMessage.Address)
+
+	return &UserProvidedData{
+		Metadata: &Claims{
+			CustomClaims: map[string]interface{}{
+				"address": parsedMessage.Address,
+				"chain":   parsedMessage.ChainID,
+				"role":    "authenticated",
+			},
+			Subject: providerId, // This becomes the provider_id in the identity
+		},
+		Emails: []Email{},
+	}, nil
+}
+
+
+func (p *Web3Provider) verifySolanaSignature(signature string, rawMessage string, msg *siws.SIWSMessage) error {
+	// Decode base64 signature into bytes
+	sigBytes, err := base64.StdEncoding.DecodeString(string(signature))
+	if err != nil {
+		return fmt.Errorf("invalid signature encoding: %w", err)
+	}
+
+	params := siws.SIWSVerificationParams{
+		ExpectedDomain: p.config.Domain,
+		CheckTime:      true,
+		TimeDuration:   p.config.Timeout,
+	}
+
+	if err := crypto.VerifySIWS(rawMessage, sigBytes, msg, params); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *Web3Provider) GenerateSignMessage(address string, chain string, uri string) (string, error) {
+	if chain == "" {
+		chain = p.defaultChain
+	}
+
+	chainCfg, ok := p.chains[chain]
+	if !ok {
+		return "", fmt.Errorf("unsupported chain: %s", chain)
+	}
+
+	// Generate nonce for message uniqueness
+	nonce := crypto.SecureToken()
```

Comment:
^ref to comment about nonce generation varying across implementation

Reply:
ah, just saw https://github.com/supabase/auth/pull/1918#discussion_r1921109157 - still think it would be good that the 3 locations that produce a nonce value are consistent in how they derive the value and the charset/length.